py-polars/polars/internals/dataframe/frame.py

```diff
@@ -1255,144 +1173,219 @@ def __getitem__(
-            item = pli.Series("", item)  # fall through to next if isinstance
-
-        if isinstance(item, pli.Series):
-            dtype = item.dtype
-            if dtype == Utf8:
-                return self._from_pydf(self._df.select(item))
-            if dtype == UInt32:
-                return self._from_pydf(self._df.take_with_series(item._s))
-            if dtype in {UInt8, UInt16, UInt64, Int8, Int16, Int32, Int64}:
-                return self._from_pydf(
-                    self._df.take_with_series(self._pos_idxs(item, dim=0)._s)
+        if _NUMPY_AVAILABLE and isinstance(rows, np.ndarray):
+            if rows.ndim != 1:
+                raise TypeError("numpy row indexer must only have a single dimension")
+            if rows.dtype in [np.uint8, np.uint16, np.uint32]:
+                return pli.Series(rows, dtype=get_idx_type())
+            if rows.dtype in [np.int8, np.int16, np.int32, np.int64, np.uint64]:
+                return self._standardize_row_accessor(pli.Series(rows))
+        raise TypeError(f"unable to determine row indices from {rows}")
+
+    def _standardize_col_accessor(
+        self, cols: int | str | MultiColSelector
+    ) -> list[pli.Expr]:
+        """
+        Transform any column accessor to a canonical type.
+
+        The canonical type is a list of ``pl.col`` expressions.
+        """
+        if isinstance(cols, (int, str)):
+            return self._standardize_col_accessor([cols])  # type: ignore[arg-type]
+        if isinstance(cols, slice):
+            # NOTE: the only valid combinations are
+            #   - slice(int | None, int | None, int | None)
+            #   - slice(str | None, str | None, int | None)
+            start, stop = self._standardize_slice_start_stop(cols.start, cols.stop)
+            rng_indices = slice(start, stop, cols.step).indices(self.width)
+            return self._standardize_col_accessor(range(*rng_indices))
+        if isinstance(cols, range):
+            return self._standardize_col_accessor([self.columns[i] for i in cols])
+        if _NUMPY_AVAILABLE and isinstance(cols, np.ndarray):
+            if cols.ndim != 1:
+                raise TypeError(
+                    "numpy column indexer must only have a single dimension"
                 )
+            return self._standardize_col_accessor(cols.tolist())
+        if isinstance(cols, list) and cols.__len__() == 0:
+            # Guard against an empty list, i.e. no columns selected
+            return []
+        if is_bool_sequence(cols):
+            if cols.__len__() != self.columns.__len__():
+                raise ValueError("boolean mask does not match DataFrame width")
+            masked = [x for x, flag in zip(self.columns, cols) if flag]
+            return self._standardize_col_accessor(masked)
+        if is_int_sequence(cols):
+            return self._standardize_col_accessor([self.columns[i] for i in cols])
+        if is_str_sequence(cols, allow_str=False):
```

Comment:
This is the exiting condition (together with empty list), would be good to move to minimize the number of if-branches to be hit, and for clarity.

Reply:
Oh I see. I was thinking the exiting condition made sense to put at the bottom, but you're suggesting for performance reasons we would actually want this at the top?

Reply:
^ Assuming that is what you meant, I've implemented it in the latest commit

Reply:
Yes, I meant move to top, but I see I made a typo. Thank you.